File: fib.go
package main

func go_rec_fib(n int) int {
	if n <= 2 {
		return 1
	}
	return go_rec_fib(n-1) + go_rec_fib(n-2)
}

func go_tail_rec_fib(n int) int {
	var f func(int, int, int) int
	f = func(term, val, prev int) int {
		if term == 0 {
			return prev
		}
		if term == 1 {
			return val
		}
		return f(term-1, prev+val, val)
	}
	return f(n, 1, 0)
}

func go_iter_fib(n int) int {
	if n <= 2 {
		return 1
	}
	prev1 := 0
	prev2 := 1
	for i := 2; i <= n; i++ {
		prev1, prev2 = prev2, prev1+prev2
	}
	return prev2
}

var go_mem_fib = memoize(go_rec_fib)

func tramp_tail_rec_fib(term, val, prev int) tramp {
	if term == 0 {
		return func() (int, tramp) {
			return prev, nil
		}
	}
	if term == 1 {
		return func() (int, tramp) {
			return val, nil
		}
	}
	return func() (int, tramp) {
		return 0, tramp_tail_rec_fib(term-1, prev+val, val)
	}
}

func go_tramp_tail_rec_fib(n int) int {
	return trampoline(tramp_tail_rec_fib(n, 1, 0))
}

File: fib.nim
import common

proc rec_fib(n: cint): cint {.exportc.} =
  if n <= 2: result = 1
  else: result = rec_fib(n-1)+rec_fib(n-2)

proc tail_rec_fib(n: cint): cint {.exportc.} =
  proc f(term, val, prev: cint): cint =
    if term == 0: result = prev
    elif term == 1: result = val
    else: result = f(term-1,prev+val,val)
  result = f(n, 1, 0)

proc iter_fib(n: cint): cint {.exportc.} =
  if n <= 2:
    result = 1
  else:
    var prev1: cint = 0
    var prev2: cint = 1
    for i in 2..n:
      let tmp = prev1
      prev1 = prev2
      prev2 = tmp+prev2
    result = prev2

let mem_fib_internal = memoize(rec_fib)

proc mem_fib(n: cint): cint {.exportc.} =
  result = mem_fib_internal(n)

discard """
# Type tramp defines a function which returns result or next trampoline.
type
  trampres = object
    res: cint
    next: tramp
  tramp = proc(): trampres

# Function trampExec executes trampoline loop.
proc trampoline(t: tramp): cint =
  var r = t()
  while r.next != nil:
    r = r.next()
  result = r.res

proc tramp_tail_rec_fib_internal(term, val, prev: cint): tramp =
  if term == 0:
    result = proc(): trampres =
      result = trampres(res: prev, next: nil)
  elif term == 1:
    result = proc(): trampres =
      result = trampres(res: val, next: nil)
  else:
    result = proc(): trampres =
      var next = tramp_tail_rec_fib_internal(term-1, prev+val, val)
      result = trampres(res: cint(0), next: next)

proc tramp_tail_rec_fib(n: cint): cint {.exportc.} =
  result = trampoline(tramp_tail_rec_fib_internal(n, 1, 0))

echo tramp_tail_rec_fib(8)
"""

# This is a stab till code above fixed.
proc tramp_tail_rec_fib(n: cint): cint {.exportc.} =
  result = 42
